File: 轮式/框架/course_agv_icp/scripts/icp.py
#!/usr/bin/env python

import rospy
import numpy as np
import math

# structure of the nearest neighbor 
class NeighBor:
    def __init__(self):
        self.distances = []
        self.src_indices = []
        self.tar_indices = []

class ICP:
    def __init__(self):
        # max iterations
        self.max_iter = rospy.get_param('/icp/max_iter',10)
        # distance threshold for filter the matching points
        self.dis_th = rospy.get_param('/icp/dis_th',3)
        # tolerance to stop icp
        self.tolerance = rospy.get_param('/icp/tolerance',0)
        # min match
        self.min_match = rospy.get_param('/icp/min_match',2)
    
    # ICP process function
    # Waiting for Implementation 
    # return: T = (R, t), where T is 2*3, R is 2*2 and t is 2*1    
    def process(self,tar_pc,src_pc):
        A = src_pc
        B = tar_pc
        m = A.shape[1]
        # make points homogeneous, copy them to maintain the originals
        src = np.ones((m+1,A.shape[0]))
        dst = np.ones((m+1,B.shape[0]))
        src[:m,:] = np.copy(A.T)
        dst[:m,:] = np.copy(B.T)
        Transfrom_acc = np.identity(3)
        prev_error = 0.0
        iter_cnt = match_cnt = 0
        #main loop
        for i in range(self.max_iter * 2):
            # print('---iter:', i) #TODO
            if (np.NaN not in src_pc and np.NaN not in tar_pc):
                neigh_ = self.findNearest(src_pc.T, tar_pc.T)
                # src_pc_xy = np.ones((2, src_pc.shape[0]))
                src_pc_xy = np.ones((2, len(src_pc)))
                tar_chorder = np.ones((2, len(tar_pc)))
                for j in range(len(neigh_.src_indices)):
                    src_pc_xy[0:2, j] = src_pc[0:2, neigh_.src_indices[j]]
                    tar_chorder[0:2, j] = tar_pc[0:2, neigh_.tar_indices[j]]                
                Transfrom = self.getTransform(src_pc_xy.T, tar_chorder.T)
                Transfrom_acc = np.dot(Transfrom, Transfrom_acc)
                src_pc = np.dot(Transfrom, src_pc)
                mean_error = np.mean(neigh_.distances)
                if np.abs(prev_error - mean_error) < self.tolerance:
                    break
                prev_error = mean_error
            iter_cnt += 1
        return Transfrom_acc

    # find the nearest points & filter
    # return: neighbors of src and tar
    def findNearest(self,src,tar):
        neigh = NeighBor()
        min = 100.0
        index = 0
        dist_temp = 0.0
        row_src = np.size(src, 0)
        row_tar = np.size(tar, 0)
        for i in range(row_src):
            lst_src = src[i, 0:2].T
            min = 0x3f3f3f3f
            index = 0
            dist_temp = 0
            for j in range(row_tar):
                lst_tar = tar[j, 0:2].T
                dist_temp = np.hypot(lst_tar[0] - lst_src[0], lst_tar[1] - lst_src[1])
                if dist_temp < min:
                    min = dist_temp
                    index = j
    
            if min < self.dis_th:
                neigh.distances.append(min)
                neigh.src_indices.append(i)
                neigh.tar_indices.append(index)
        return neigh

    # Waiting for Implementation 
    # return: T = (R, t), where T is 2*3, R is 2*2 and t is 2*1
    def getTransform(self,src,tar):
        T = np.identity(3)
        num = src.shape[0]
        #归一化
        centroid_src = np.mean(src, axis = 0)
        centroid_tar = np.mean(tar, axis = 0)
        tar_ = tar - centroid_tar
        src_ = src - centroid_src
        H = np.dot(src_.T, tar_)
        U, S, Vt = np.linalg.svd(H)
        R = np.dot(Vt.T, U.T)
        t = centroid_tar - np.dot(R, centroid_src)
        T[:2, :2] = R
        T[:2, 2] = t
        return T

    def calcDist(self,a,b):
        dx = a[0] - b[0]
        dy = a[1] - b[1]
        return math.hypot(dx,dy)

File: 轮式/框架/course_agv_icp/src/icp.cpp
#include "ros/ros.h"
#include "ros/console.h"
#include <stdio.h>

#include <numeric>
#include <vector>
#include <Eigen/Eigen>

#include "ros/publisher.h"
#include "ros/subscriber.h"
#include "nav_msgs/Odometry.h"
#include <tf/transform_broadcaster.h>
#include <sensor_msgs/LaserScan.h>

using namespace std;
using namespace Eigen;

// structure of the nearest neighbor 
typedef struct{
    std::vector<float> distances;
    std::vector<int> src_indices;
    std::vector<int> tar_indices;
} NeighBor;

class icp{

public:

    icp(ros::NodeHandle &n);
    ~icp();
    ros::NodeHandle& n;

    // robot init states
    double robot_x;
    double robot_y;
    double robot_theta;
    // sensor states = robot_x_y_theta
    Vector3d sensor_sta;

    // max iterations
    int max_iter;
    // distance threshold for filter the matching points
    double dis_th;
    // tolerance to stop icp
    double tolerance;
    // if is the first scan, set as the map/target
    bool isFirstScan;
    // src point cloud matrix
    MatrixXd src_pc;
    // target point cloud matrix
    MatrixXd tar_pc;

    // ICP process function
    void process(sensor_msgs::LaserScan input);
    // transform the ros msg to Eigen Matrix
    Eigen::MatrixXd rosmsgToEigen(const sensor_msgs::LaserScan input);
    // fint the nearest points & filter
    NeighBor findNearest(const Eigen::MatrixXd &src, const Eigen::MatrixXd &tar);
    // get the transform from two point sets in one iteration
    Eigen::Matrix3d getTransform(const Eigen::MatrixXd &A, const Eigen::MatrixXd &B);
    // calc 2D Euclidean distance
    float calc_dist(const Eigen::Vector2d &pta, const Eigen::Vector2d &ptb);
    // transform vector states to matrix form
    Eigen::Matrix3d staToMatrix(const Vector3d sta);
  
    // ros-related subscribers, publishers and broadcasters
    ros::Subscriber laser_sub;
    void publishResult(Matrix3d T);
 	tf::TransformBroadcaster odom_broadcaster;
 	ros::Publisher odom_pub;
};

icp::~icp()
{}

icp::icp(ros::NodeHandle& n):
    n(n)
{	

	// get the params
	n.getParam("/icp/robot_x", robot_x);
	n.getParam("/icp/robot_y", robot_y);
	n.getParam("/icp/robot_theta", robot_theta);
	sensor_sta << robot_x, robot_y, robot_theta;

	n.getParam("/icp/max_iter", max_iter);
	n.getParam("/icp/tolerance", tolerance);
	n.getParam("/icp/dis_th", dis_th);

    isFirstScan = true;
    laser_sub = n.subscribe("/scan", 1, &icp::process, this);
    odom_pub = n.advertise<nav_msgs::Odometry>("icp_odom", 1);
}

void icp::process(sensor_msgs::LaserScan input)
{
    cout<<"------seq:  "<<input.header.seq<<endl;

    // set the inital
    if(isFirstScan)
    {
        tar_pc = this->rosmsgToEigen(input);
        isFirstScan =false;
        return;
    }

	double time_0 = (double)ros::Time::now().toSec();

    // init some variables
    NeighBor neigh_;
    src_pc = this->rosmsgToEigen(input);
    cout<<"input cnt:  "<<src_pc.cols()<<endl;

    Eigen::MatrixXd src_pc_xy;
    Eigen::MatrixXd tar_chorder;

    Eigen::Matrix3d Transform;
    Eigen::Matrix3d Transform_acc = Eigen::MatrixXd::Identity(3,3);
    double prev_error = 0;
    double mean_error;
    int iter_cnt = 0;

    int match_cnt = 0;

    // main LOOP
    for(int i=0; i<max_iter; i++)
    {	
    	// cout<<"---iter: "<<i<<endl;
        neigh_ = this->findNearest(src_pc.transpose(), tar_pc.transpose());

    	src_pc_xy = Eigen::MatrixXd::Ones(2, neigh_.src_indices.size());
    	tar_chorder = Eigen::MatrixXd::Ones(2, neigh_.src_indices.size());
        
        // filter ...
        for(int j=0; j<neigh_.src_indices.size(); j++)
        {	
        		src_pc_xy.block<2,1>(0,j) = src_pc.block<2,1>(0, neigh_.src_indices[j]);
	            tar_chorder.block<2,1>(0,j) = tar_pc.block<2,1>(0, neigh_.tar_indices[j]);
        }
        // filter finished ...

        Transform = this->getTransform(src_pc_xy.transpose(), tar_chorder.transpose());

        Transform_acc = Transform * Transform_acc;

        src_pc = Transform * src_pc;

        mean_error = std::accumulate(neigh_.distances.begin(),neigh_.distances.end(),0.0)/neigh_.distances.size();

        if (abs(prev_error - mean_error) < tolerance)
            break;

        prev_error = mean_error;
    	iter_cnt ++;
    }
    cout << "total_iter: " << iter_cnt << endl;

    tar_pc = this->rosmsgToEigen(input);

    this->publishResult(Transform_acc);

	double time_1 = (double)ros::Time::now().toSec();
	cout<<"time_cost:  "<<time_1-time_0<<endl;
}

Eigen::MatrixXd icp::rosmsgToEigen(const sensor_msgs::LaserScan input)
{
    int total_num = (input.angle_max - input.angle_min) / input.angle_increment + 1;

    Eigen::MatrixXd pc = Eigen::MatrixXd::Ones(3,total_num);

    float angle;
    for(int i=0; i<total_num; i++)
    {
        angle = input.angle_min + i * input.angle_increment;

        pc(0,i) = input.ranges[i] * std::cos(angle);
        pc(1,i) = input.ranges[i] * std::sin(angle);
    }
    return pc;
}

NeighBor icp::findNearest(const Eigen::MatrixXd &src, const Eigen::MatrixXd &tar)
{
    // temp cpy
    int row_src = src.rows();
    int row_tar = tar.rows();
    Eigen::Vector2d vec_src;
    Eigen::Vector2d vec_tar;

    NeighBor neigh;
    float min = 100;
    int index = 0;
    float dist_temp = 0;

    for(int ii=0; ii < row_src; ii++){
        vec_src = src.block<1,2>(ii,0).transpose();
        min = 100;
        index = 0;
        dist_temp = 0;
        for(int jj=0; jj < row_tar; jj++){
            vec_tar = tar.block<1,2>(jj,0).transpose();
            dist_temp = calc_dist(vec_src,vec_tar);
            if (dist_temp < min){
                min = dist_temp;
                index = jj;
            }
        }

        //filter added
        if(min < dis_th)
        {
        	neigh.distances.push_back(min);
        	neigh.src_indices.push_back(ii);
        	neigh.tar_indices.push_back(index);
        }
    }
    return neigh;
}

Eigen::Matrix3d icp::getTransform(const Eigen::MatrixXd &src, const Eigen::MatrixXd &tar)
{
    Eigen::Matrix3d T = Eigen::MatrixXd::Identity(3, 3);
    Eigen::Vector2d centroid_src(0,0);
    Eigen::Vector2d centroid_tar(0,0);
    Eigen::MatrixXd src_ = src;
    Eigen::MatrixXd tar_ = tar;
    int num = src.rows();

    for(int i=0; i<num; i++){
        centroid_src += src.block<1,2>(i,0).transpose();
        centroid_tar += tar.block<1,2>(i,0).transpose();
    }
    centroid_src /= num;
    centroid_tar /= num;

    for(int i=0; i<num; i++){
        src_.block<1,2>(i,0) = src.block<1,2>(i,0) - centroid_src.transpose();
        tar_.block<1,2>(i,0) = tar.block<1,2>(i,0) - centroid_tar.transpose();
    }

    Eigen::MatrixXd H = src_.transpose()*tar_;
    Eigen::MatrixXd U;
    Eigen::VectorXd S;
    Eigen::MatrixXd V;
    Eigen::MatrixXd Vt;
    Eigen::Matrix2d R;
    Eigen::Vector2d t;

    JacobiSVD<Eigen::MatrixXd> svd(H, ComputeFullU | ComputeFullV);
    U = svd.matrixU();
    S = svd.singularValues();
    V = svd.matrixV();
    Vt = V.transpose();

    R = Vt.transpose()*U.transpose();

    t = centroid_tar - R*centroid_src;

    T.block<2,2>(0,0) = R;
    T.block<2,1>(0,2) = t;
    return T;
}

float icp::calc_dist(const Eigen::Vector2d &pta, const Eigen::Vector2d &ptb)
{
    return std::sqrt((pta[0]-ptb[0])*(pta[0]-ptb[0]) + (pta[1]-ptb[1])*(pta[1]-ptb[1]));
}

Eigen::Matrix3d icp::staToMatrix(Eigen::Vector3d sta)
{
	Matrix3d RT;
    RT << cos(sta(2)), -sin(sta(2)), sta(0),
          sin(sta(2)), cos(sta(2)),sta(1),
          0, 0, 1;
    return RT;
}

void icp::publishResult(Eigen::Matrix3d T)
{	
    float delta_yaw = atan2(T(1,0), T(0,0));
    cout<<"sensor-delta-xyt: "<<T(0,2)<<" "<<T(1,2)<<" "<<delta_yaw<<endl;

    sensor_sta(0) = sensor_sta(0) + cos(sensor_sta(2))*T(0,2) - sin(sensor_sta(2))*T(1,2);
    sensor_sta(1) = sensor_sta(1) + sin(sensor_sta(2))*T(0,2) + cos(sensor_sta(2))*T(1,2);
    sensor_sta(2) = sensor_sta(2) + delta_yaw;

    cout<<"sensor-global: "<<sensor_sta.transpose()<<endl;

    // tf
    geometry_msgs::Quaternion odom_quat = tf::createQuaternionMsgFromYaw(sensor_sta(2));

    geometry_msgs::TransformStamped odom_trans;
    odom_trans.header.stamp = ros::Time::now();
    odom_trans.header.frame_id = "world_base";
    odom_trans.child_frame_id = "icp_odom";

    odom_trans.transform.translation.x = sensor_sta(0);
    odom_trans.transform.translation.y = sensor_sta(1);
    odom_trans.transform.translation.z = 0.0;
    odom_trans.transform.rotation = odom_quat;

    odom_broadcaster.sendTransform(odom_trans);

    // odom
    nav_msgs::Odometry odom;
    odom.header.stamp = ros::Time::now();
    odom.header.frame_id = "world_base";

    odom.pose.pose.position.x = sensor_sta(0);
    odom.pose.pose.position.y = sensor_sta(1);
    odom.pose.pose.position.z = 0.0;
    odom.pose.pose.orientation = odom_quat;

    odom_pub.publish(odom);
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "icp");
    ros::NodeHandle n;

    icp icp_(n);

    ros::MultiThreadedSpinner spinner(1);
    spinner.spin();

    return 0;
}